mesonbuild/modules/python.py

```diff
@@ -0,0 +1,384 @@
+# Copyright 2018 The Meson development team
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import os
+
+from pathlib import Path
+from .. import mesonlib
+from . import ExtensionModule
+from mesonbuild.modules import ModuleReturnValue
+from . import permittedSnippetKwargs
+from ..interpreterbase import (
+    noPosargs, noKwargs, permittedKwargs,
+    InterpreterObject, InvalidArguments
+)
+from ..interpreter import shlib_kwargs, ExternalProgramHolder
+from .. import mlog
+from ..environment import detect_cpu_family
+from ..dependencies.base import (
+    DependencyMethods, ExternalDependency,
+    ExternalProgram, PkgConfigDependency,
+    NonExistingExternalProgram
+)
+
+mod_kwargs = set(['subdir'])
+mod_kwargs.update(shlib_kwargs)
+mod_kwargs -= set(['install_dir', 'install_rpath'])
+
+
+def run_command(python, command):
+    _, stdout, _ = mesonlib.Popen_safe(python.get_command() + [
+        '-c',
+        command])
+
+    return stdout.strip()
+
+
+class PythonDependency(ExternalDependency):
+    def __init__(self, python_holder, environment, kwargs):
+        super().__init__('python', environment, None, kwargs)
+        self.name = 'python'
+        self.static = kwargs.get('static', False)
+        self.version = python_holder.version
+        self.platform = python_holder.platform
+        self.pkgdep = None
+        self.variables = python_holder.variables
+        self.paths = python_holder.paths
+        if mesonlib.version_compare(self.version, '>= 3.0'):
+            self.major_version = 3
+        else:
+            self.major_version = 2
+
+        if DependencyMethods.PKGCONFIG in self.methods:
```

Comment:
As much as i like using pkg-config for everything, i think this is asking for trouble with a custom python executable.
Maybe we could add a check that at least tries to see if the pkg-config result matches the python executable?
Also at least debian installs .pc files with the full python version (i.e. python-2.7.pc, python-3.5.pc, python-3.5m.pc) which should be tried before generic names.

Not sure what might be an more specific alternative, there seems to be python${FOO}-config files. But they also seem to output different data than the .pc files.

Reply:
This is taken straight from https://github.com/mesonbuild/meson/blob/master/mesonbuild/dependencies/misc.py#L262 fwiw :)

Reply:
Yes, that dependency had the semantic of finding 'some' python.
While here the python that should satisfy this dependency is potentially specified by absolute path. So if the user has more control, there is a greater expectation that that control actually does what it says on the tin.

Reply:
Looking at how to improve this, it's valid criticism, I just think at least some of the improvements will have to come with usage :)

Reply:
```
from distutils import sysconfig
print(sysconfig.get_config_var("CFLAGS"))
print(sysconfig.get_config_var("LDFLAGS"))
```

this is what disutils does: https://github.com/python/cpython/blob/dfb6e54dd8dbd735f55109ad8ee9dfcb6178ede9/Lib/distutils/sysconfig.py#L150

Reply:
I think pkg-config works, just need to perform a bit more validation as @textshell suggests

Reply:
On the other hand, distutils might be the canonical definition of the expected behavior for building python modules. But it looks like it might do a lot more than what meson wants (i.e. it's own environment variable based overrides)

Reply:
let me tr things, not relying on pkg-config would indeed be nice

Reply:
It turns out that the path to the pkg-config directory is exposed in the sysconfig variables, this in addition with the version number lets us use pkg-config in a way that I think is now safe and portable, what do you think @textshell ?

Reply:
I tend to not using pkg-config as primary dependency provider if we can sanely get the information directly from the python interpreter.
But if we can make sure that we only pick up pkg-config file with the complete version from the directory that the python interpreter says it's supposed to be installed that should work also.

Reply:
this is what the code now does, and it seems reasonably portable to me, if in the future this proves not to be we can always change that, as it is not part of the interface, for what it's worth this module has now been tested on OSX (not by me), multiple distributions of Linux and Windows, with both python 2 and 3 :)